File: tools/upload_server.py
#!/usr/bin/env python3
"""AIM 上传服务：音频（/）+ 视频（/video）"""
import os
import sys
import html
import re
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import cgi

SAVE_DIR = "/root/aim/audio_upload"
VIDEO_DIR = "/tmp/aim_video_upload"
os.makedirs(SAVE_DIR, exist_ok=True)
os.makedirs(VIDEO_DIR, exist_ok=True)

PAGE = """<!DOCTYPE html>
<html lang="zh">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>AIM 音频上传</title>
<style>
  body { background:#11110f; color:#fff5dc; font-family:monospace; padding:24px; max-width:560px; margin:0 auto; }
  h1 { color:#ff4e35; font-size:20px; }
  h2 { color:#ffd36a; font-size:14px; margin-top:28px; }
  .box { background:#1e1d1a; border:1px solid #5a554c; padding:16px; margin-bottom:12px; }
  label { display:block; margin-bottom:6px; color:#77736b; font-size:12px; }
  input[type=file] { color:#fff5dc; margin-bottom:10px; width:100%; }
  input[type=text] { background:#11110f; border:1px solid #5a554c; color:#fff5dc; padding:6px; width:100%; box-sizing:border-box; margin-bottom:10px; }
  button { background:#ff4e35; border:none; color:#fff; padding:10px 20px; font-size:14px; cursor:pointer; font-family:monospace; }
  button:hover { filter:brightness(1.15); }
  .files { list-style:none; padding:0; }
  .files li { padding:6px 0; border-bottom:1px solid #2a2824; font-size:13px; }
  .ok { color:#61d39e; margin-top:10px; font-size:13px; }
  .err { color:#ff4e35; margin-top:10px; font-size:13px; }
  .note { color:#77736b; font-size:11px; margin-top:16px; }
</style>
</head>
<body>
<h1>AIM 音频上传</h1>
<div class="box">
  <h2>🎵 非战斗 BGM（数码闲时）</h2>
  <label>选择文件（mp3/wav/ogg）</label>
  <input type="file" id="f1" accept="audio/*">
  <label>保存为文件名（默认 bgm_idle.mp3）</label>
  <input type="text" id="n1" value="bgm_idle.mp3">
  <button onclick="up(1)">上传</button>
  <div id="s1"></div>
</div>
<div class="box">
  <h2>⚔️ 战斗 BGM（数码指挥部）</h2>
  <label>选择文件（mp3/wav/ogg）</label>
  <input type="file" id="f2" accept="audio/*">
  <label>保存为文件名（默认 bgm_battle.mp3）</label>
  <input type="text" id="n2" value="bgm_battle.mp3">
  <button onclick="up(2)">上传</button>
  <div id="s2"></div>
</div>
<h2>已接收文件</h2>
<ul class="files" id="flist">__FILES__</ul>
<p class="note">上传后显示 ✅ 即成功。文件会保存到服务器，随后集成进游戏。</p>
<script>
function up(n) {
  var file = document.getElementById('f'+n).files[0];
  var name = document.getElementById('n'+n).value.trim() || ('bgm_'+(n==1?'idle':'battle')+'.mp3');
  var box = document.getElementById('s'+n);
  if (!file) { box.innerHTML = '<div class="err">请先选择文件</div>'; return; }
  box.innerHTML = '<div class="ok">上传中…</div>';
  var xhr = new XMLHttpRequest();
  xhr.open('POST', '/upload?name=' + encodeURIComponent(name), true);
  xhr.onload = function() {
    if (xhr.status == 200) {
      box.innerHTML = '<div class="ok">✅ 上传成功：' + name + '</div>';
      setTimeout(function(){ location.reload(); }, 1200);
    } else {
      box.innerHTML = '<div class="err">上传失败：' + xhr.responseText + '</div>';
    }
  };
  xhr.onerror = function() { box.innerHTML = '<div class="err">网络错误</div>'; };
  xhr.send(file);
}
</script>
</body>
</html>"""

VIDEO_PAGE = """<!DOCTYPE html>
<html lang="zh">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>AIM 视频上传</title>
<style>
body { background:#11110f; color:#fff5dc; font-family:monospace; padding:24px; max-width:600px; margin:0 auto; }
h1 { color:#ff4e35; font-size:20px; }
h2 { color:#ffd36a; font-size:14px; margin-top:28px; }
.box { background:#1e1d1a; border:1px solid #5a554c; padding:16px; margin-bottom:12px; }
label { display:block; margin-bottom:6px; color:#77736b; font-size:12px; }
input[type=file] { color:#fff5dc; margin-bottom:10px; width:100%; }
button { background:#ff4e35; color:#fff; border:2px solid #5a554c; padding:10px 20px; font-size:14px; cursor:pointer; }
.ok { color:#5ac87a; font-size:13px; }
.err { color:#ff6b6b; font-size:13px; }
ul { list-style:none; padding:0; }
li { padding:6px 10px; background:#1e1d1a; border:1px solid #5a554c; margin-bottom:4px; font-size:13px; }
li a { color:#ffd36a; text-decoration:none; }
.sz { color:#77736b; font-size:11px; margin-left:8px; }
.nav { color:#77736b; font-size:12px; margin-bottom:16px; }
.nav a { color:#ffd36a; text-decoration:none; }
</style>
</head>
<body>
<h1>AIM 视频上传</h1>
<div class="nav"><a href="/audio">→ 音频上传页</a></div>
<p style="color:#77736b; font-size:13px;">选择游戏视频上传，我下载下来分析后告诉你哪里要改。建议10MB 以内。</p>
<div class="box">
<form id="vf" action="/video" method="post" enctype="multipart/form-data">
<label>选择视频文件（MP4 / MOV）</label>
<input type="file" name="video" id="vfile" accept="video/*" required>
<br>
<button type="submit">上传</button>
</form>
<div id="vstat"></div>
</div>
<h2>已上传视频</h2>
<ul>__FILES__</ul>
<script>
document.getElementById('vf').addEventListener('submit', function(e) {
  e.preventDefault();
  var f = document.getElementById('vfile').files[0];
  if (!f) return;
  document.getElementById('vstat').innerHTML = '<div class="ok">上传中…</div>';
  var fd = new FormData();
  fd.append('video', f);
  var xhr = new XMLHttpRequest();
  xhr.open('POST', '/video', true);
  xhr.onload = function() { if (xhr.status < 400) location.reload(); else document.getElementById('vstat').innerHTML = '<div class="err">上传失败</div>'; };
  xhr.send(fd);
});
</script>
</body>
</html>"""

class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        u = urlparse(self.path)
        # 游戏介绍：HTML 展示页 + TXT 下载
        if u.path == "/aim_intro" or u.path == "/aim_intro.html":
            hp = os.path.join(os.path.dirname(os.path.abspath(__file__)), "aim_intro.html")
            if os.path.isfile(hp):
                with open(hp, "rb") as f:
                    data = f.read()
                self.send_response(200)
                self.send_header("Content-Type", "text/html; charset=utf-8")
                self.send_header("Content-Length", str(len(data)))
                self.end_headers()
                self.wfile.write(data)
                return
            self.send_response(404); self.end_headers(); return
        # 游戏介绍 TXT 下载
        if u.path == "/aim_intro.txt":
            p = os.path.join(os.path.dirname(os.path.abspath(__file__)), "aim_intro.txt")
            if os.path.isfile(p):
                with open(p, "rb") as f:
                    data = f.read()
                self.send_response(200)
                self.send_header("Content-Type", "text/plain; charset=utf-8")
                self.send_header("Content-Length", str(len(data)))
                self.send_header("Content-Disposition", "attachment; filename=aim_intro.txt")
                self.end_headers()
                self.wfile.write(data)
                return
            self.send_response(404); self.end_headers(); return
        if u.path == "/" or u.path == "/index.html":
            self.send_response(303); self.send_header("Location", "/audio"); self.end_headers()
            return
        if u.path == "/audio" or u.path == "/audio.html":
            files = sorted(os.listdir(SAVE_DIR)) if os.path.isdir(SAVE_DIR) else []
            items = ""
            for f in files:
                p = os.path.join(SAVE_DIR, f)
                sz = os.path.getsize(p)
                kb = sz / 1024
                items += f"<li>{html.escape(f)} — {kb:.1f} KB</li>"
            page = PAGE.replace("__FILES__", items)
            data = page.encode()
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.send_header("Content-Length", str(len(data)))
            self.end_headers()
            self.wfile.write(data)
            return
        if u.path == "/video" or u.path == "/video.html":
            files = []
            for f in sorted(os.listdir(VIDEO_DIR)):
                p = os.path.join(VIDEO_DIR, f)
                if os.path.isfile(p):
                    sz = os.path.getsize(p)
                    files.append((f, sz))
            items = ""
            for name, sz in files:
                mb = sz / (1024*1024)
                items += f'<li><a href="/videofile/{html.escape(name)}">{html.escape(name)}</a> <span class="sz">{mb:.2f} MB</span></li>'
            page = VIDEO_PAGE.replace("__FILES__", items or '<li style="color:#77736b">还没有视频</li>')
            data = page.encode()
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.send_header("Content-Length", str(len(data)))
            self.end_headers()
            self.wfile.write(data)
            return
        if u.path.startswith("/videofile/"):
            fname = os.path.basename(u.path[len("/videofile/"):])
            p = os.path.join(VIDEO_DIR, fname)
            if os.path.isfile(p):
                sz = os.path.getsize(p)
                ext = os.path.splitext(fname)[1].lower()
                mime = "video/mp4" if ext == ".mp4" else "video/quicktime" if ext == ".mov" else "application/octet-stream"
                self.send_response(200)
                self.send_header("Content-Type", mime)
                self.send_header("Content-Length", str(sz))
                self.send_header("Cache-Control", "no-store")
                self.end_headers()
                with open(p, "rb") as f:
                    while chunk := f.read(64*1024):
                        self.wfile.write(chunk)
                return
            self.send_response(404); self.end_headers(); return
        self.send_response(404)
        self.end_headers()

    def do_POST(self):
        u = urlparse(self.path)
        # 视频上传（/video）
        if u.path == "/video":
            n = int(self.headers.get("Content-Length", 0))
            data = self.rfile.read(n)
            m = re.search(rb'filename="([^"]+)"', data)
            if not m:
                self.send_response(400); self.end_headers(); return
            fname = os.path.basename(m.group(1).decode("utf-8", errors="replace"))
            boundary = self.headers.get("Content-Type", "").split("boundary=")[-1].encode()
            parts = data.split(b"--" + boundary)
            for p in parts:
                if b"Content-Type:" in p and b"\r\n\r\n" in p:
                    head, body = p.split(b"\r\n\r\n", 1)
                    if body.endswith(b"\r\n"):
                        body = body[:-2]
                    if fname and len(body) > 1024:
                        out = os.path.join(VIDEO_DIR, fname)
                        with open(out, "wb") as f:
                            f.write(body)
                        print(f"[video] {fname} ({len(body)/1024/1024:.2f} MB) -> {out}", flush=True)
                        break
            self.send_response(303); self.send_header("Location", "/video"); self.end_headers()
            return
        if u.path == "/video-list":
            self.send_response(303); self.send_header("Location", "/video"); self.end_headers()
            return
        if u.path != "/upload":
            self.send_response(404)
            self.end_headers()
            return
        qs = parse_qs(u.query)
        name = (qs.get("name") or [""])[0].strip()
        if not name:
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b"missing name")
            return
        name = os.path.basename(name)
        ctype = self.headers.get("Content-Type", "")
        if "multipart/form-data" in ctype:
            # 浏览器 XHR send(file) 是原始二进制；form 表单则走 multipart
            form = cgi.FieldStorage(
                fp=self.rfile,
                headers=self.headers,
                environ={"REQUEST_METHOD": "POST", "CONTENT_TYPE": ctype})
            body = b""
            if form and form.file:
                body = form.file.read()
        else:
            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length) if length > 0 else b""
        if not body:
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b"empty body")
            return
        path = os.path.join(SAVE_DIR, name)
        with open(path, "wb") as f:
            f.write(body)
        self.send_response(200)
        self.send_header("Content-Type", "text/plain; charset=utf-8")
        self.end_headers()
        self.wfile.write(f"OK saved {name} {len(body)} bytes".encode())

    def log_message(self, *a):
        pass

if __name__ == "__main__":
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 8001
    srv = HTTPServer(("0.0.0.0", port), Handler)
    print(f"upload server on :{port}, save to {SAVE_DIR}")
    srv.serve_forever()

File: server/src/upload_server.js
// AIM 素材接收服务 — 独立端口 8001
// 用途：牢大把收款码 / Logo 传上来：
//   POST /upload        → public/downloads/wechat_qr.png
//   POST /upload_logo   → public/downloads/logo.png
// 支持两种上传方式：
//   1. multipart/form-data:  curl -F "file=@qr.png" http://<host>:8001/upload
//   2. JSON base64:          curl -H "Content-Type: application/json" -d '{"base64":"..."}' http://<host>:8001/upload
// GET /  -> 网页上传页（拖拽/选择 + 预览 + 当前生效图）
'use strict';
const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = 8001;
const SAVE_DIR = path.join(__dirname, '..', 'public', 'downloads');
const TARGETS = {
  upload: { file: path.join(SAVE_DIR, 'wechat_qr.png'), key: 'qr' },
  upload_logo: { file: path.join(SAVE_DIR, 'logo.png'), key: 'logo' },
  upload_image: { file: path.join(SAVE_DIR, 'aim_upload.png'), key: 'gen' }, // 通用图片（B 站素材/截图等，离离读图用）
};

fs.mkdirSync(SAVE_DIR, { recursive: true });

function sendJson(res, code, obj) {
  const body = JSON.stringify(obj);
  res.writeHead(code, { 'Content-Type': 'application/json; charset=utf-8' });
  res.end(body);
}

// 解析 multipart/form-data，取出第一个文件字段的二进制
function parseMultipart(buf, contentType) {
  const m = /boundary=(?:"([^"]+)"|([^;]+))/i.exec(contentType || '');
  if (!m) return null;
  const boundary = '--' + (m[1] || m[2]);
  const parts = buf.toString('binary').split(boundary);
  for (const part of parts) {
    if (!part || part.trim() === '--') continue;
    if (part.startsWith('--')) continue;
    const headerEnd = part.indexOf('\r\n\r\n');
    if (headerEnd < 0) continue;
    const header = part.slice(0, headerEnd);
    const bodyBin = part.slice(headerEnd + 4);
    const data = bodyBin.replace(/\r\n$/, '');
    if (!/name="/i.test(header)) continue;
    const nameM = /name="([^"]+)"/i.exec(header);
    if (!nameM) continue;
    const fieldName = nameM[1];
    if (fieldName === 'file' || fieldName === 'image' || fieldName === 'qr' || fieldName === 'logo') {
      return Buffer.from(data, 'binary');
    }
  }
  return null;
}

function saveImage(res, buf, target, okMsg) {
  if (!buf || buf.length === 0) {
    sendJson(res, 400, { ok: false, msg: '没解析到图片，请用 -F "file=@xx.png" 或 JSON base64' });
    return;
  }
  if (fs.existsSync(target.file)) {
    try { fs.copyFileSync(target.file, target.file + '.bak'); } catch (e) { /* ignore */ }
  }
  fs.writeFileSync(target.file, buf);
  sendJson(res, 200, { ok: true, msg: okMsg, size: buf.length, url: path.basename(target.file) });
}

function handleUpload(req, res, target, okMsg) {
  const chunks = [];
  let size = 0;
  req.on('data', c => {
    size += c.length;
    if (size > 10 * 1024 * 1024) {
      req.destroy();
      sendJson(res, 413, { ok: false, msg: '图片太大（上限 10MB）' });
      return;
    }
    chunks.push(c);
  });
  req.on('end', () => {
    const buf = Buffer.concat(chunks);
    const ctype = req.headers['content-type'] || '';
    let image = null;
    if (ctype.startsWith('multipart/form-data')) {
      image = parseMultipart(buf, ctype);
    } else if (ctype.includes('application/json')) {
      try {
        const obj = JSON.parse(buf.toString('utf8'));
        image = obj.base64 ? Buffer.from(obj.base64, 'base64') : null;
      } catch (e) { /* fallthrough */ }
    }
    if (!image || image.length === 0) {
      // 表单提交失败 → 跳回页面提示
      if (req.url.includes('form=1')) {
        res.writeHead(302, { Location: '/?err=parse' });
        res.end();
        return;
      }
      sendJson(res, 400, { ok: false, msg: '没解析到图片，请用 -F "file=@xx.png" 或 JSON base64' });
      return;
    }
    if (fs.existsSync(target.file)) {
      try { fs.copyFileSync(target.file, target.file + '.bak'); } catch (e) { /* ignore */ }
    }
    fs.writeFileSync(target.file, image);
    if (req.url.includes('form=1')) {
      res.writeHead(302, { Location: '/?ok=1' });
      res.end();
      return;
    }
    sendJson(res, 200, { ok: true, msg: okMsg, size: image.length, url: path.basename(target.file) });
  });
}

// 网页上传页（两个上传卡片：收款码 / Logo）
function pageHtml(qrExists, logoExists, genExists, host) {
  return `<!DOCTYPE html>
<html lang="zh">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>AIM · 素材上传</title>
<style>
  * { margin: 0; padding: 0; box-sizing: border-box; }
  body {
    background: #0d1020; color: #e8e8ff;
    font-family: 'Courier New', monospace;
    min-height: 100vh; display: flex; align-items: center; justify-content: center;
    padding: 16px; image-rendering: pixelated;
  }
  .row { display: flex; gap: 16px; flex-wrap: wrap; justify-content: center; }
  .card {
    background: #141a30; border: 3px solid #3a4a7a;
    box-shadow: 0 0 0 3px #0d1020, 0 0 0 6px #3a4a7a, 0 8px 0 rgba(0,0,0,.5);
    padding: 24px; width: 100%; max-width: 400px; text-align: center;
  }
  h1 { font-size: 18px; letter-spacing: 2px; color: #ffd75e; margin-bottom: 4px; text-shadow: 2px 2px 0 #000; }
  .sub { font-size: 12px; color: #8fa0d0; margin-bottom: 16px; }
  .drop {
    border: 3px dashed #3a4a7a; background: #0f1530; padding: 22px 10px;
    cursor: pointer; transition: all .15s; font-size: 13px; color: #aab8e8;
    position: relative; overflow: hidden;
  }
  .legacy { margin-top: 18px; padding-top: 14px; border-top: 1px dashed #3a4a7a; font-size: 12px; color: #8fa0d0; }
  .legacy input[type=file] { color: #aab8e8; font-size: 12px; margin-bottom: 8px; }
  .legacy button {
    background: #3a6ea5; color: #fff; border: 2px solid #6aa8e0;
    padding: 8px 18px; font-family: inherit; font-size: 13px; cursor: pointer;
  }
  .drop:hover, .drop.over { border-color: #ffd75e; background: #161d3a; color: #ffd75e; }
  .preview {
    max-width: 220px; max-height: 160px; margin: 14px auto; display: none;
    image-rendering: pixelated; border: 3px solid #3a4a7a; background: #000;
  }
  .preview.show { display: block; }
  .btn {
    display: block; width: 100%; margin-top: 12px; padding: 13px;
    font-family: inherit; font-size: 15px; letter-spacing: 3px;
    background: #3a6ea5; color: #fff; border: 3px solid #6aa8e0; border-bottom-width: 6px;
    cursor: pointer; text-shadow: 1px 1px 0 #000;
  }
  .btn:hover { background: #4a7eb5; }
  .btn:active { transform: translateY(3px); border-bottom-width: 3px; }
  .btn:disabled { background: #2a3a5a; border-color: #4a5a7a; cursor: not-allowed; }
  .status { margin-top: 12px; font-size: 13px; min-height: 20px; }
  .ok { color: #6fe06f; } .err { color: #ff7070; }
  .cur { margin-top: 10px; font-size: 12px; color: #8fa0d0; }
  .cur img { max-width: 140px; max-height: 100px; margin-top: 8px; border: 2px solid #3a4a7a; background: #000; }
</style>
</head>
<body>
<div class="row">
  ${card('✦ 投喂收款码 ✦', '微信收款码，投喂按钮用的', 'upload', 'wechat_qr', qrExists, 'qr')}
  ${card('✦ 开机 Logo ✦', '透明底白字 logo，启动页淡入用', 'upload_logo', 'logo', logoExists, 'logo')}
  ${card('✦ 通用图片 ✦', '任意截图/素材（离离读取分析）', 'upload_image', 'gen', genExists, 'gen')}
</div>
<script>
  function pick(input, preview, up, status, file) {
    if (file.size > 10 * 1024 * 1024) { status.className = 'status err'; status.textContent = '图片太大（上限 10MB）'; return; }
    input._f = file;
    try { preview.src = URL.createObjectURL(file); } catch (e) { preview.src = ''; }
    preview.classList.add('show');
    up.disabled = false;
    status.className = 'status';
    status.textContent = '已选：' + file.name + '（' + Math.round(file.size / 1024) + ' KB）';
  }
  function uploadNow(target, input, up, status, curId) {
    if (!input._f) return;
    up.disabled = true; status.className = 'status'; status.textContent = '传送中…';
    const fd = new FormData();
    fd.append('file', input._f);
    fetch('/' + target, { method: 'POST', body: fd })
      .then(r => r.json())
      .then(d => {
        if (d.ok) {
          status.className = 'status ok';
          status.textContent = '✅ 成功！已生效';
          document.getElementById(curId).innerHTML = '<img src="/preview_' + target + '?t=' + Date.now() + '">';
        } else {
          status.className = 'status err';
          status.textContent = '失败：' + (d.msg || '未知错误');
          up.disabled = false;
        }
      })
      .catch(e => { status.className = 'status err'; status.textContent = '网络错误：' + e; up.disabled = false; });
  }
  function bind(target, previewId, upId, statusId, curId) {
    const drop = document.getElementById(target + '_drop');
    const input = document.getElementById(target + '_file');
    const preview = document.getElementById(previewId);
    const up = document.getElementById(upId);
    const status = document.getElementById(statusId);
    // input 透明铺满 drop，点击/选择都是原生行为；拖拽事件绑在 input 上（冒泡到 drop）
    input.ondragover = e => { e.preventDefault(); drop.classList.add('over'); };
    input.ondragleave = () => drop.classList.remove('over');
    input.ondrop = e => { e.preventDefault(); drop.classList.remove('over'); if (e.dataTransfer.files[0]) pick(input, preview, up, status, e.dataTransfer.files[0]); };
    // 选完文件立即自动上传（不需要点任何按钮）
    input.onchange = () => {
      if (input.files && input.files[0]) {
        pick(input, preview, up, status, input.files[0]);
        uploadNow(target, input, up, status, curId);
      }
    };
    // 「传送」按钮：自动上传失败时手动重试
    up.onclick = () => uploadNow(target, input, up, status, curId);
  }
  bind('upload', 'qr_preview', 'qr_up', 'qr_status', 'qr_cur');
  bind('upload_logo', 'logo_preview', 'logo_up', 'logo_status', 'logo_cur');
  bind('upload_image', 'gen_preview', 'gen_up', 'gen_status', 'gen_cur');
</script>
</body>
</html>`;

  function card(title, sub, target, key, exists, curId) {
    const cur = exists ? '已上传' : '还没有';
    return `<div class="card">
  <h1>${title}</h1>
  <div class="sub">${sub}</div>
  <div class="drop" id="${target}_drop">
    点击选择图片<br>或直接拖进来<br><br>
    <span style="color:#6aa8e0">支持 png / jpg</span>
    <input type="file" id="${target}_file" accept="image/*"
           style="position:absolute; inset:0; width:100%; height:100%; opacity:0; cursor:pointer;">
  </div>
  <img class="preview" id="${key}_preview" alt="预览">
  <button class="btn" id="${key}_up" disabled>传 送</button>
  <div class="status" id="${key}_status">当前：${cur}</div>
  <div class="cur" id="${key}_cur">${exists ? '<img src="/preview_' + target + '?t=' + Date.now() + '">' : ''}</div>
  <div class="legacy">📎 如果上面自动上传没反应，用这个经典方式：
    <form action="/${target}?form=1" method="post" enctype="multipart/form-data">
      <input type="file" name="file" accept="image/*">
      <button type="submit">选择并上传</button>
    </form>
  </div>
</div>`;
  }
}

const server = http.createServer((req, res) => {
  const url = (req.url || '/').split('?')[0];
  console.log(`[AIM-upload] ${new Date().toISOString()} ${req.method} ${req.url} from ${req.socket.remoteAddress}`);
  res.on('finish', () => console.log(`  → ${res.statusCode}`));

  if (req.method === 'GET') {
    const qrExists = fs.existsSync(TARGETS.upload.file);
    const logoExists = fs.existsSync(TARGETS.upload_logo.file);
    const genExists = fs.existsSync(TARGETS.upload_image.file);
    if (url === '/preview_upload') {
      if (!qrExists) { sendJson(res, 404, { ok: false, msg: '还没有收款码' }); return; }
      res.writeHead(200, { 'Content-Type': 'image/png', 'Cache-Control': 'no-store' });
      fs.createReadStream(TARGETS.upload.file).pipe(res);
      return;
    }
    if (url === '/preview_upload_logo') {
      if (!logoExists) { sendJson(res, 404, { ok: false, msg: '还没有 Logo' }); return; }
      res.writeHead(200, { 'Content-Type': 'image/png', 'Cache-Control': 'no-store' });
      fs.createReadStream(TARGETS.upload_logo.file).pipe(res);
      return;
    }
    if (url === '/preview_upload_image') {
      if (!genExists) { sendJson(res, 404, { ok: false, msg: '还没有图片' }); return; }
      res.writeHead(200, { 'Content-Type': 'image/png', 'Cache-Control': 'no-store' });
      fs.createReadStream(TARGETS.upload_image.file).pipe(res);
      return;
    }
    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
    res.end(pageHtml(qrExists, logoExists, genExists, req.headers.host));
    return;
  }

  if (req.method === 'POST') {
    if (url === '/upload') {
      handleUpload(req, res, TARGETS.upload, '收款码已更新');
      return;
    }
    if (url === '/upload_logo') {
      handleUpload(req, res, TARGETS.upload_logo, 'Logo 已更新');
      return;
    }
    if (url === '/upload_image') {
      handleUpload(req, res, TARGETS.upload_image, '图片已更新');
      return;
    }
  }

  sendJson(res, 404, { ok: false, msg: 'not found' });
});

server.listen(PORT, () => {
  console.log(`[AIM-upload] 素材接收端口 ${PORT} 已启动`);
  console.log(`  GET  ${PORT}/                  上传页（收款码 + Logo）`);
  console.log(`  POST ${PORT}/upload             收款码 → downloads/wechat_qr.png`);
  console.log(`  POST ${PORT}/upload_logo        Logo   → downloads/logo.png`);
});
